### PR TITLE
Add cherry-pick-approved prow plugin

### PR DESCRIPTION
**Pull request by saschagrunert**
The plugin can be used by Release Managers to automatically manage the labels for cherry-picks by submitting GitHub reviews.

Fixes: https://github.com/kubernetes/sig-release/issues/2083

cc @kubernetes/release-engineering 

**Before I extend the tests, please review the overall logic and provide feedback.**

### CODE REVIEW

**Review by cpanato — APPROVED** [review]
looks cool, thank you!

/lgtm

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:39]: I think it might be better if we hard-code this to "cherry-pick-approved", we do the same in the [cherry-pick-unapproved plugin](https://github.com/kubernetes/test-infra/blob/63b8f838e6fcb32cc81b3d753cbc763f6ea336ae/prow/plugins/cherrypickunapproved/cherrypick-unapproved.go#L39). If the label name eventually gets changed in the future, it might break Prow which is an unintended side-effect

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:45]: This should be configurable via flag or config file (whichever is possible for built-in plugin). This plugin might also be used by projects other than Kubernetes, and those projects might use some different branch regexp (e.g. `release/v*`)

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:55]: On similar note as the previous comment, I recommend updating description to abstract away Kubernetes Release Managers. Maybe something like `The %s plugin approves cherry-pick to be merged when approved by reviewers authorized for cherry-picks`

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:120]: I think that we should return an error here instead of just logging it.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:128]: Should we include here some information like the PR number?

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:135]: I strongly recommend removing the first two checks. As I commented earlier, this plugin might also be used by projects other than Kubernetes.

```suggestion
	if !branchRegexp.MatchString(branch) {
```

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:136]: This shouldn't error, instead it should return nil

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:141]: This shouldn't error, instead it should return nil

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:148]: This shouldn't error, instead it should return nil

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:167]: LGTM shouldn't be an invalid label (unless I get this wrong)

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:193]: This is also done by cherry-pick-unapproved plugin, so we have to make sure we'll have no conflicts between two plugins

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:182]: Is it possible to check here if all tests are passing? If it is, I think we should add it to the criteria, because we tend to cherry-pick-approve PRs only if tests are passing

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:200]: The URL should be configurable

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:202]: The error message should be generalized (i.e. without mentioning k/k)

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (xmudrii) [inline prow/plugins/cherrypickapproved/cherrypickapproved.go:216]: The release managers alias name should be configurable

**Review by xmudrii — CHANGES_REQUESTED** [review]
I'm really happy to see this plugin being integrated into Prow, thanks a lot @saschagrunert for working on this! It's definitely going to help the Kubernetes Release Managers a lot!

The PR looks solid, but given that this plugin is being integrated into Prow, I think that it should be generalized as much as possible so that other Prow users can benefit from it (similar as for the cherry-pick-unapproved plugin). Most of comments are related to that and the error handling.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:39]: Took this inspiration from the LGTM plugin, but I can change the const to text-only.

https://github.com/kubernetes/test-infra/blob/36a7c91fffc75a3bc9dd8d18a6a6471e7b3a9527/prow/plugins/lgtm/lgtm.go#L38

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:120]: That's tricky since I'm not exactly sure how prow behaves here. Anyways, changing to return the `err`.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:45]: Done :+1:

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:55]: Fixed

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:128]: Good idea, added `org`, `repo`, `branch` and `pr`.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:135]: Yes, then we should probably return nil here and skip it.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:136]: Yep, done.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:141]: Done

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:148]: Done

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:167]: That's true, removed it.

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go:182]: Yes, I will add that

File: prow/plugins/cherrypickapproved/cherrypickapproved.go
Comment (saschagrunert) [reply prow/plugins/cherrypickapproved/cherrypickapproved.go]: Done.

File: config/prow/plugins.yaml
Comment (xmudrii) [inline config/prow/plugins.yaml:1107]: I have two concerns about this:
- This is pulling from GitHub which is prone to errors and flakes (e.g. rate limits, `raw.githubusercontent.com` being unavailable)
- The original idea behind this plugin was to allow more broader group to approve cherry-picks. Concretely, we could allow Release Manager Associates that are good at reviewing cherry-picks to approve them before becoming fully-fledged Release Managers. I think that would be appreciated by both RMs and RMAs: RMs would have help from RMAs, RMAs would have more tasks to work on

That said, it might be better if we have explicit list of folks allowed to approve cherry-picks.

I see two risks with this though:
- We need to keep the list up to date manually, but given that the Release Management team is not changed too often and that we have onboarding/offboarding checklists -- I don't see this being a problem
- We need to make sure that someone who's not part of the Release Management team doesn't add themselves to the list. I think that Prow config changes should be already strict, but this is the risk that we have to keep in mind

Comment (saschagrunert) [reply config/prow/plugins.yaml]: I don't agree with all of the points, but generally I'd agree that a list might be safer to parse. Switching it over to a manually maintained list.

**Comment by saschagrunert** [comment]
Added unit tests. PTAL again @kubernetes/release-managers

File: config/prow/plugins.yaml
Comment (xmudrii) [inline config/prow/plugins.yaml:1107]: Nit: let's leave a comment with URL where this list is coming from (e.g. OWNERS_ALIASES in k/k or [k/website](https://kubernetes.io/releases/release-managers/#release-managers))

**Review by xmudrii — APPROVED** [review]
One additional nit: it might be useful if this plugin could leave a comment on PR if approval is not successful (e.g. test is failing) to give a signal to the release manager that another approval will be needed.

Other than that, this is amazing, thank you for this, @saschagrunert! I'd recommend as the next step to share this PR in #prow and #sig-testing Slack channels to get awareness from folks who can approve this PR.

/lgtm
/approve

File: config/prow/plugins.yaml
Comment (saschagrunert) [reply config/prow/plugins.yaml]: Yep, sure thing. Added a link.

**Review by cpanato — APPROVED** [review]
thanks that is nice

/lgtm

**Pull request merged**
